File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/banner/BannerEditViewModel.kt
package com.hhp227.concafe.presentation.banner

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.common.AppError
import com.hhp227.concafe.domain.model.BannerLinkTargetType
import com.hhp227.concafe.domain.model.HomeBannerCreate
import com.hhp227.concafe.domain.model.UserRole
import com.hhp227.concafe.domain.usecase.CreateHomeBannerUseCase
import com.hhp227.concafe.domain.usecase.GetCafeEventPageUseCase
import com.hhp227.concafe.domain.usecase.GetCafeManagementUseCase
import com.hhp227.concafe.domain.usecase.GetCafeNoticePageUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class BannerEditViewModel(
    private val initialCafeId: String? = null,
    private val createHomeBannerUseCase: CreateHomeBannerUseCase,
    private val getCafeManagementUseCase: GetCafeManagementUseCase,
    private val getCafeNoticePageUseCase: GetCafeNoticePageUseCase,
    private val getCafeEventPageUseCase: GetCafeEventPageUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(BannerEditUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<BannerEditEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private var selectorJob: Job? = null

    fun onAction(action: BannerEditAction) {
        when (action) {
            BannerEditAction.ClickBack -> clickBack()
            BannerEditAction.ClickImagePicker -> clickImagePicker()
            is BannerEditAction.ChangeTitle -> _uiState.update { it.copy(title = action.value) }
            is BannerEditAction.ChangeSubtitle -> _uiState.update { it.copy(subtitle = action.value) }
            is BannerEditAction.SelectTarget -> selectTarget(action.target)
            is BannerEditAction.ChangeTargetValue -> _uiState.update { it.copy(targetValue = action.value) }
            is BannerEditAction.ChangeDisplayDays -> _uiState.update {
                it.copy(displayDays = action.value.coerceIn(1, 10))
            }
            BannerEditAction.ClickCafeSelector -> openCafeSelector()
            BannerEditAction.ClickTargetSelector -> openTargetSelector()
            is BannerEditAction.ChangeSelectorQuery -> changeSelectorQuery(action.value)
            is BannerEditAction.SelectSelectorItem -> selectSelectorItem(action.id)
            BannerEditAction.DismissSelector -> dismissSelector()
            BannerEditAction.ClickSave -> clickSave()
            BannerEditAction.DismissInfoMessage -> _uiState.update { it.copy(infoMessage = null) }
        }
    }

    private fun loadOwnedCafeOptions() {
        viewModelScope.launch {
            when (val result = getCafeManagementUseCase.invoke()) {
                is AppResult.Success -> {
                    val options = result.data.ownedCafes.map { cafe ->
                        BannerSelectableItem(
                            id = cafe.id,
                            title = cafe.name,
                            subtitle = cafe.city
                        )
                    }
                    _uiState.update { state ->
                        val selectedCafe = initialCafeId?.let { targetCafeId ->
                            options.firstOrNull { it.id == targetCafeId }
                        } ?: state.selectedCafeOption?.let { current ->
                            options.firstOrNull { it.id == current.id }
                        } ?: if (state.isAdmin) {
                            state.selectedCafeOption
                        } else {
                            options.firstOrNull()
                        }
                        state.copy(
                            ownedCafeOptions = options,
                            selectedCafeOption = selectedCafe,
                            selectedContentOption = if (selectedCafe?.id == state.selectedCafeOption?.id) {
                                state.selectedContentOption
                            } else {
                                null
                            },
                            targetValue = when (state.selectedTarget) {
                                BannerTargetType.CAFE_DETAIL -> selectedCafe?.id.orEmpty()
                                BannerTargetType.EXTERNAL_LINK -> state.targetValue
                                else -> if (selectedCafe?.id == state.selectedCafeOption?.id) state.targetValue else ""
                            }
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            ownedCafeOptions = emptyList(),
                            infoMessage = "운영 카페 목록을 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun observeSession() {
        viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest { user ->
                _uiState.update { it.copy(isAdmin = user?.role == UserRole.ADMIN) }
            }
        }
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(BannerEditEvent.NavigateBack)
        }
    }

    private fun clickImagePicker() {
        _uiState.update {
            it.copy(
                selectedImageLabel = "banner_cover_mock.png",
                infoMessage = "이미지 업로드 연결은 다음 단계에서 구현됩니다."
            )
        }
    }

    private fun selectTarget(target: BannerTargetType) {
        _uiState.update { state ->
            state.copy(
                selectedTarget = target,
                selectedCafeOption = when (target) {
                    BannerTargetType.CAFE_DETAIL -> when {
                        state.isAdmin -> state.selectedCafeOption
                        initialCafeId != null -> state.ownedCafeOptions.firstOrNull { it.id == initialCafeId }
                        else -> state.selectedCafeOption ?: state.ownedCafeOptions.firstOrNull()
                    }
                    else -> state.selectedCafeOption
                },
                targetValue = when (target) {
                    BannerTargetType.CAFE_DETAIL -> when {
                        state.isAdmin -> state.selectedCafeOption?.id.orEmpty()
                        initialCafeId != null -> state.ownedCafeOptions.firstOrNull { it.id == initialCafeId }?.id.orEmpty()
                        else -> (state.selectedCafeOption ?: state.ownedCafeOptions.firstOrNull())?.id.orEmpty()
                    }
                    BannerTargetType.EXTERNAL_LINK -> ""
                    else -> ""
                },
                selectedContentOption = null,
                selectorType = null,
                selectorQuery = "",
                selectorOptions = emptyList(),
                isSelectorLoading = false
            )
        }
    }

    private fun openCafeSelector() {
        _uiState.update { state ->
            state.copy(
                selectorType = BannerSelectorType.CAFE,
                selectorQuery = "",
                selectorOptions = state.ownedCafeOptions,
                isSelectorLoading = false
            )
        }
    }

    private fun openTargetSelector() {
        val currentState = _uiState.value
        val selectedCafe = currentState.selectedCafeOption

        if (selectedCafe == null) {
            _uiState.update { it.copy(infoMessage = "먼저 운영 카페를 선택해주세요.") }
            return
        }

        when (currentState.selectedTarget) {
            BannerTargetType.NOTICE -> {
                _uiState.update {
                    it.copy(
                        selectorType = BannerSelectorType.NOTICE,
                        selectorQuery = "",
                        selectorOptions = emptyList(),
                        isSelectorLoading = true
                    )
                }
                loadNoticeOptions(selectedCafe.id, query = "")
            }
            BannerTargetType.EVENT_DETAIL -> {
                _uiState.update {
                    it.copy(
                        selectorType = BannerSelectorType.EVENT,
                        selectorQuery = "",
                        selectorOptions = emptyList(),
                        isSelectorLoading = true
                    )
                }
                loadEventOptions(selectedCafe.id, query = "")
            }
            else -> Unit
        }
    }

    private fun changeSelectorQuery(value: String) {
        _uiState.update { it.copy(selectorQuery = value) }
        when (_uiState.value.selectorType) {
            BannerSelectorType.CAFE -> {
                _uiState.update { state ->
                    state.copy(selectorOptions = state.ownedCafeOptions.filter {
                        it.title.contains(value, ignoreCase = true) ||
                            it.subtitle.contains(value, ignoreCase = true)
                    })
                }
            }
            BannerSelectorType.NOTICE -> {
                _uiState.value.selectedCafeOption?.id?.let { loadNoticeOptions(it, value) }
            }
            BannerSelectorType.EVENT -> {
                _uiState.value.selectedCafeOption?.id?.let { loadEventOptions(it, value) }
            }
            null -> Unit
        }
    }

    private fun loadNoticeOptions(cafeId: String, query: String) {
        selectorJob?.cancel()
        selectorJob = viewModelScope.launch {
            _uiState.update { it.copy(isSelectorLoading = true) }
            when (val result = getCafeNoticePageUseCase.invoke(cafeId, query, cursor = null, pageSize = 50)) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            selectorOptions = result.data.items.map { item ->
                                BannerSelectableItem(
                                    id = item.id,
                                    title = item.title,
                                    subtitle = item.displayDate
                                )
                            },
                            isSelectorLoading = false
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            selectorOptions = emptyList(),
                            isSelectorLoading = false,
                            infoMessage = "공지사항 목록을 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun loadEventOptions(cafeId: String, query: String) {
        selectorJob?.cancel()
        selectorJob = viewModelScope.launch {
            _uiState.update { it.copy(isSelectorLoading = true) }
            when (val result = getCafeEventPageUseCase.invoke(cafeId, query, cursor = null, pageSize = 50)) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            selectorOptions = result.data.items.map { item ->
                                BannerSelectableItem(
                                    id = item.id,
                                    title = item.title,
                                    subtitle = item.periodText
                                )
                            },
                            isSelectorLoading = false
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            selectorOptions = emptyList(),
                            isSelectorLoading = false,
                            infoMessage = "이벤트 목록을 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun selectSelectorItem(id: String) {
        val currentState = _uiState.value
        when (currentState.selectorType) {
            BannerSelectorType.CAFE -> {
                val selectedCafe = currentState.ownedCafeOptions.firstOrNull { it.id == id } ?: return
                val isSameCafe = currentState.selectedCafeOption?.id == selectedCafe.id
                _uiState.update {
                    it.copy(
                        selectedCafeOption = selectedCafe,
                        selectedContentOption = if (isSameCafe) it.selectedContentOption else null,
                        targetValue = when (it.selectedTarget) {
                            BannerTargetType.CAFE_DETAIL -> selectedCafe.id
                            BannerTargetType.EXTERNAL_LINK -> it.targetValue
                            else -> if (isSameCafe) it.targetValue else ""
                        },
                        selectorType = null,
                        selectorQuery = "",
                        selectorOptions = emptyList(),
                        isSelectorLoading = false
                    )
                }
            }
            BannerSelectorType.NOTICE,
            BannerSelectorType.EVENT -> {
                val selectedTargetItem = currentState.selectorOptions.firstOrNull { it.id == id } ?: return
                _uiState.update {
                    it.copy(
                        selectedContentOption = selectedTargetItem,
                        targetValue = selectedTargetItem.id,
                        selectorType = null,
                        selectorQuery = "",
                        selectorOptions = emptyList(),
                        isSelectorLoading = false
                    )
                }
            }
            null -> Unit
        }
    }

    private fun dismissSelector() {
        selectorJob?.cancel()
        _uiState.update {
            it.copy(
                selectorType = null,
                selectorQuery = "",
                selectorOptions = emptyList(),
                isSelectorLoading = false
            )
        }
    }

    private fun clickSave() {
        val currentState = _uiState.value

        val validationMessage = when {
            currentState.title.isBlank() -> "배너 제목을 입력해주세요."
            currentState.subtitle.isBlank() -> "서브 문구를 입력해주세요."
            currentState.selectedTarget == BannerTargetType.EXTERNAL_LINK &&
                currentState.targetValue.isBlank() -> "외부 URL을 입력해주세요."
            currentState.selectedTarget != BannerTargetType.EXTERNAL_LINK &&
                currentState.targetValue.isBlank() -> "연결 대상을 선택해주세요."
            else -> null
        }

        if (validationMessage != null) {
            _uiState.update { it.copy(infoMessage = validationMessage) }
            return
        }

        _uiState.update { it.copy(isSaving = true, infoMessage = null) }
        viewModelScope.launch {
            when (val result = createHomeBannerUseCase.invoke(currentState.toCreateInput())) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            infoMessage = "배너가 등록되었습니다."
                        )
                    }
                    _event.emit(BannerEditEvent.ShowSaveSuccessMessage)
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            infoMessage = result.error.toUserMessage()
                        )
                    }
                }
            }
        }
    }

    init {
        observeSession()
        loadOwnedCafeOptions()
    }
}

private fun BannerEditUiState.toCreateInput(): HomeBannerCreate {
    return HomeBannerCreate(
        cafeId = selectedCafeOption?.id,
        title = title.trim(),
        subtitle = subtitle.trim(),
        imageUrl = selectedImageLabel,
        targetType = selectedTarget.toDomainType(),
        targetValue = targetValue.trim(),
        displayDays = displayDays
    )
}

private fun BannerTargetType.toDomainType(): BannerLinkTargetType {
    return when (this) {
        BannerTargetType.CAFE_DETAIL -> BannerLinkTargetType.CAFE_DETAIL
        BannerTargetType.EVENT_DETAIL -> BannerLinkTargetType.EVENT_DETAIL
        BannerTargetType.NOTICE -> BannerLinkTargetType.NOTICE
        BannerTargetType.EXTERNAL_LINK -> BannerLinkTargetType.EXTERNAL_LINK
    }
}

private fun AppError.toUserMessage(): String {
    return when (this) {
        AppError.Unauthorized -> "로그인 후 배너를 등록해주세요."
        AppError.PermissionDenied -> "배너 등록 권한이 없습니다."
        AppError.NotFound -> "연결 대상을 찾을 수 없습니다."
        is AppError.ValidationFailed -> reason
        is AppError.NetworkError -> "배너를 등록하지 못했습니다."
        is AppError.Unknown -> "배너 등록 중 오류가 발생했습니다."
    }
}

File: iosApp/iosApp/Presentation/Main/CafeManagement/BannerEdit/BannerEditViewModel.swift
//
//  BannerEditViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/14.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class BannerEditViewModel: ObservableObject {
    private let initialCafeId: String?

    private let initialBannerId: String?

    private let createHomeBannerUseCase: CreateHomeBannerUseCase

    private let updateHomeBannerUseCase: UpdateHomeBannerUseCase

    private let getCafeManagementUseCase: GetCafeManagementUseCase

    private let getHomeBannerManagementUseCase: GetHomeBannerManagementUseCase

    private let getCafeNoticePageUseCase: GetCafeNoticePageUseCase

    private let getCafeEventPageUseCase: GetCafeEventPageUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let uploadImageUseCase: UploadImageUseCase

    @Published private(set) var uiState = BannerEditUiState()

    let event = PassthroughSubject<BannerEditEvent, Never>()

    private var selectorTask: Task<Void, Never>?

    private func observeSession() {
        Task {
            do {
                for try await user in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.uiState.isAdmin = user?.role == .admin
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func loadOwnedCafeOptions() {
        Task {
            do {
                let result = try await getCafeManagementUseCase.invoke()
                if let success = result as? AppResultSuccess<AnyObject>,
                   let data = success.data as? CafeManagementData {
                    let options = data.ownedCafes
                    uiState.ownedCafeOptions = options
                    if let initialCafeId {
                        uiState.selectedCafeId = options.first(where: { $0.id == initialCafeId })?.id
                    } else if uiState.selectedCafeOption == nil && !uiState.isAdmin {
                        uiState.selectedCafeId = options.first?.id
                    }
                    if uiState.selectedTarget == .cafeDetail {
                        uiState.targetValue = uiState.selectedCafeOption?.id ?? ""
                    }
                } else {
                    uiState.ownedCafeOptions = []
                    uiState.infoMessage = MessageKey.ownedCafeLoadFailed
                }
            } catch {
                uiState.ownedCafeOptions = []
                uiState.infoMessage = MessageKey.ownedCafeLoadFailed
            }
        }
    }

    private func loadEditingBanner() {
        guard let initialBannerId else { return }
        Task {
            do {
                let result = try await getHomeBannerManagementUseCase.invoke(cafeId: initialCafeId)
                if let success = result as? AppResultSuccess<AnyObject>,
                   let banners = success.data as? [HomeBanner] {
                    if let banner = banners.first(where: { $0.id == initialBannerId }) {
                        applyEditingBanner(banner)
                    } else {
                        uiState.infoMessage = MessageKey.editBannerNotFound
                    }
                } else {
                    uiState.infoMessage = MessageKey.editBannerLoadFailed
                }
            } catch {
                uiState.infoMessage = MessageKey.editBannerLoadFailed
            }
        }
    }

    private func applyEditingBanner(_ banner: HomeBanner) {
        let target: BannerTargetType
        switch banner.targetType {
        case .cafeDetail:
            target = .cafeDetail
        case .eventDetail:
            target = .eventDetail
        case .notice:
            target = .notice
        case .externalLink:
            target = .externalLink
        default:
            target = .externalLink
        }

        uiState.editingBannerId = banner.id
        uiState.selectedImageLabel = banner.imageUrl
        uiState.originalImageUrl = banner.imageUrl
        uiState.title = banner.title
        uiState.subtitle = banner.subtitle
        uiState.displayDays = min(max(Int(banner.displayDays), 1), 10)
        uiState.selectedCafeId = banner.cafeId
        uiState.selectedTarget = target
        uiState.targetValue = banner.targetValue
        uiState.selectedNoticeId = uiState.selectedTarget == .notice ? banner.targetValue : nil
        uiState.selectedEventId = uiState.selectedTarget == .eventDetail ? banner.targetValue : nil
        uiState.infoMessage = nil

        if let cafeId = banner.cafeId, !cafeId.isEmpty {
            if uiState.selectedTarget == .notice {
                loadNoticeOptions(cafeId: cafeId, query: "")
            } else if uiState.selectedTarget == .eventDetail {
                loadEventOptions(cafeId: cafeId, query: "")
            }
        }
    }

    private func selectTarget(_ target: BannerTargetType) {
        uiState.selectedTarget = target
        uiState.selectedNoticeId = nil
        uiState.selectedEventId = nil
        uiState.selectorType = nil
        uiState.selectorQuery = ""
        uiState.noticeSelectorOptions = []
        uiState.eventSelectorOptions = []
        uiState.isSelectorLoading = false
        switch target {
        case .cafeDetail:
            if let initialCafeId {
                uiState.selectedCafeId = uiState.ownedCafeOptions.first(where: { $0.id == initialCafeId })?.id
            } else if uiState.selectedCafeOption == nil && !uiState.isAdmin {
                uiState.selectedCafeId = uiState.ownedCafeOptions.first?.id
            }
            uiState.targetValue = uiState.selectedCafeOption?.id ?? ""
        case .externalLink:
            uiState.targetValue = ""
        case .notice, .eventDetail:
            uiState.targetValue = ""
        }
    }

    private func openTargetSelector() {
        guard let selectedCafe = uiState.selectedCafeOption else {
            uiState.infoMessage = MessageKey.selectCafeFirst
            return
        }
        switch uiState.selectedTarget {
        case .notice:
            uiState.selectorType = .notice
            uiState.selectorQuery = ""
            uiState.noticeSelectorOptions = []
            uiState.isSelectorLoading = true
            loadNoticeOptions(cafeId: selectedCafe.id, query: "")
        case .eventDetail:
            uiState.selectorType = .event
            uiState.selectorQuery = ""
            uiState.eventSelectorOptions = []
            uiState.isSelectorLoading = true
            loadEventOptions(cafeId: selectedCafe.id, query: "")
        default:
            break
        }
    }

    private func changeSelectorQuery(_ value: String) {
        uiState.selectorQuery = value
        switch uiState.selectorType {
        case .cafe:
            break
        case .notice:
            if let cafeId = uiState.selectedCafeOption?.id {
                loadNoticeOptions(cafeId: cafeId, query: value)
            }
        case .event:
            if let cafeId = uiState.selectedCafeOption?.id {
                loadEventOptions(cafeId: cafeId, query: value)
            }
        case .none:
            break
        }
    }

    private func loadNoticeOptions(cafeId: String, query: String) {
        selectorTask?.cancel()
        selectorTask = Task {
            do {
                let result = try await getCafeNoticePageUseCase.invoke(cafeId: cafeId, query: query, cursor: nil, pageSize: 50)
                if let success = result as? AppResultSuccess<AnyObject>,
                   let page = success.data as? PagedResult<CafeNoticeManagementItem> {
                    let items = page.items as! [CafeNoticeManagementItem]
                    uiState.noticeSelectorOptions = items
                    uiState.isSelectorLoading = false
                } else {
                    uiState.noticeSelectorOptions = []
                    uiState.isSelectorLoading = false
                    uiState.infoMessage = MessageKey.noticeListLoadFailed
                }
            } catch {
                if Task.isCancelled { return }
                uiState.noticeSelectorOptions = []
                uiState.isSelectorLoading = false
                uiState.infoMessage = MessageKey.noticeListLoadFailed
            }
        }
    }

    private func loadEventOptions(cafeId: String, query: String) {
        selectorTask?.cancel()
        selectorTask = Task {
            do {
                let result = try await getCafeEventPageUseCase.invoke(cafeId: cafeId, query: query, cursor: nil, pageSize: 50)
                if let success = result as? AppResultSuccess<AnyObject>,
                   let page = success.data as? PagedResult<CafeEventManagementItem> {
                    let items = page.items as! [CafeEventManagementItem]
                    uiState.eventSelectorOptions = items
                    uiState.isSelectorLoading = false
                } else {
                    uiState.eventSelectorOptions = []
                    uiState.isSelectorLoading = false
                    uiState.infoMessage = MessageKey.eventListLoadFailed
                }
            } catch {
                if Task.isCancelled { return }
                uiState.eventSelectorOptions = []
                uiState.isSelectorLoading = false
                uiState.infoMessage = MessageKey.eventListLoadFailed
            }
        }
    }

    private func selectSelectorItem(_ id: String) {
        switch uiState.selectorType {
        case .cafe:
            guard let selectedCafe = uiState.ownedCafeOptions.first(where: { $0.id == id }) else { return }
            let sameCafe = uiState.selectedCafeId == selectedCafe.id
            uiState.selectedCafeId = selectedCafe.id
            if uiState.selectedTarget == .cafeDetail {
                uiState.targetValue = selectedCafe.id
            } else if !sameCafe {
                uiState.selectedNoticeId = nil
                uiState.selectedEventId = nil
                uiState.targetValue = ""
            }
        case .notice:
            guard let selectedItem = uiState.noticeSelectorOptions.first(where: { $0.id == id }) else { return }
            uiState.selectedNoticeId = selectedItem.id
            uiState.selectedEventId = nil
            uiState.targetValue = selectedItem.id
        case .event:
            guard let selectedItem = uiState.eventSelectorOptions.first(where: { $0.id == id }) else { return }
            uiState.selectedEventId = selectedItem.id
            uiState.selectedNoticeId = nil
            uiState.targetValue = selectedItem.id
        case .none:
            return
        }
        dismissSelector()
    }

    private func dismissSelector() {
        selectorTask?.cancel()
        uiState.selectorType = nil
        uiState.selectorQuery = ""
        uiState.isSelectorLoading = false
    }

    private func clickSave() {
        let validationMessage: String?

        if uiState.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            validationMessage = MessageKey.validationTitleRequired
        } else if uiState.selectedImageLabel?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true {
            validationMessage = MessageKey.validationImageRequired
        } else if uiState.subtitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            validationMessage = MessageKey.validationSubtitleRequired
        } else if uiState.selectedTarget == .externalLink &&
                    uiState.targetValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            validationMessage = MessageKey.validationExternalUrlRequired
        } else if uiState.selectedTarget != .externalLink &&
                    uiState.targetValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            validationMessage = MessageKey.validationTargetRequired
        } else {
            validationMessage = nil
        }

        if let validationMessage {
            if uiState.selectedImageLabel?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true {
                uiState.isImageRequiredAlertVisible = true
            } else {
                uiState.infoMessage = validationMessage
            }
            return
        }

        let targetType: BannerLinkTargetType
        switch uiState.selectedTarget {
        case .cafeDetail:
            targetType = .cafeDetail
        case .eventDetail:
            targetType = .eventDetail
        case .notice:
            targetType = .notice
        case .externalLink:
            targetType = .externalLink
        }

        uiState.isSaving = true
        uiState.infoMessage = nil
        Task {
            do {
                guard let imageUrl = try await resolveBannerImageUrl() else {
                    uiState.isSaving = false
                    uiState.infoMessage = MessageKey.imageUploadFailed
                    return
                }

                let createInput = HomeBannerCreate(
                    cafeId: uiState.selectedCafeId,
                    title: uiState.title.trimmingCharacters(in: .whitespacesAndNewlines),
                    subtitle: uiState.subtitle.trimmingCharacters(in: .whitespacesAndNewlines),
                    imageUrl: imageUrl,
                    targetType: targetType,
                    targetValue: uiState.targetValue.trimmingCharacters(in: .whitespacesAndNewlines),
                    displayDays: Int32(uiState.displayDays)
                )
                let result: Any
                if let editingBannerId = uiState.editingBannerId, !editingBannerId.isEmpty {
                    result = try await updateHomeBannerUseCase.invoke(
                        bannerId: editingBannerId,
                        input: createInput
                    )
                } else {
                    result = try await createHomeBannerUseCase.invoke(input: createInput)
                }
                if result is AppResultSuccess<AnyObject> {
                    uiState.isSaving = false
                    uiState.infoMessage = nil
                    event.send(.navigateBack)
                } else if let failure = result as? AppResultFailure {
                    let userMessage: String
                    if failure.error is AppErrorUnauthorized {
                        userMessage = MessageKey.unauthorized
                    } else if failure.error is AppErrorPermissionDenied {
                        userMessage = MessageKey.permissionDenied
                    } else if failure.error is AppErrorNotFound {
                        userMessage = MessageKey.targetNotFound
                    } else if let error = failure.error as? AppErrorValidationFailed {
                        userMessage = error.reason
                    } else if failure.error is AppErrorNetworkError {
                        userMessage = MessageKey.networkFailed
                    } else {
                        userMessage = MessageKey.saveUnknown
                    }
                    uiState.isSaving = false
                    uiState.infoMessage = userMessage
                } else {
                    uiState.isSaving = false
                    uiState.infoMessage = MessageKey.saveUnknown
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isSaving = false
                uiState.infoMessage = MessageKey.saveUnknown
            }
        }
    }

    private func resolveBannerImageUrl() async throws -> String? {
        let selectedImage = uiState.selectedImageLabel?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if selectedImage.isEmpty {
            return nil
        }

        if selectedImage.hasPrefix("http://") || selectedImage.hasPrefix("https://") {
            return selectedImage
        }
        if let originalImageUrl = uiState.originalImageUrl, originalImageUrl == selectedImage {
            return originalImageUrl
        }

        let uploadResult = try await uploadImageUseCase.invoke(
            localPath: selectedImage,
            folder: "banners"
        )
        if let uploadSuccess = uploadResult as? AppResultSuccess<AnyObject>,
           let uploadedImageUrl = uploadSuccess.data as? String {
            return uploadedImageUrl
        }
        return nil
    }
    
    func onAction(_ action: BannerEditAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .clickImagePicker:
            uiState.infoMessage = nil
        case .selectImage(let imageUrl):
            uiState.selectedImageLabel = imageUrl
            uiState.infoMessage = nil
            uiState.isImageRequiredAlertVisible = false
        case .changeTitle(let value):
            uiState.title = value
        case .changeSubtitle(let value):
            uiState.subtitle = value
        case .selectTarget(let target):
            selectTarget(target)
        case .changeTargetValue(let value):
            uiState.targetValue = value
        case .changeDisplayDays(let value):
            uiState.displayDays = min(max(value, 1), 10)
        case .clickCafeSelector:
            uiState.selectorType = .cafe
            uiState.selectorQuery = ""
            uiState.isSelectorLoading = false
        case .clickTargetSelector:
            openTargetSelector()
        case .changeSelectorQuery(let value):
            changeSelectorQuery(value)
        case .selectSelectorItem(let id):
            selectSelectorItem(id)
        case .dismissSelector:
            dismissSelector()
        case .dismissImageRequiredAlert:
            uiState.isImageRequiredAlertVisible = false
        case .clickSave:
            clickSave()
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    init(
        initialCafeId: String? = nil,
        initialBannerId: String? = nil,
        createHomeBannerUseCase: CreateHomeBannerUseCase = KoinInitializerKt.resolveCreateHomeBannerUseCase(),
        updateHomeBannerUseCase: UpdateHomeBannerUseCase = KoinInitializerKt.resolveUpdateHomeBannerUseCase(),
        getCafeManagementUseCase: GetCafeManagementUseCase = KoinInitializerKt.resolveGetCafeManagementUseCase(),
        getHomeBannerManagementUseCase: GetHomeBannerManagementUseCase = KoinInitializerKt.resolveGetHomeBannerManagementUseCase(),
        getCafeNoticePageUseCase: GetCafeNoticePageUseCase = KoinInitializerKt.resolveGetCafeNoticePageUseCase(),
        getCafeEventPageUseCase: GetCafeEventPageUseCase = KoinInitializerKt.resolveGetCafeEventPageUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        uploadImageUseCase: UploadImageUseCase = KoinInitializerKt.resolveUploadImageUseCase()
    ) {
        self.initialCafeId = initialCafeId
        self.initialBannerId = initialBannerId
        self.createHomeBannerUseCase = createHomeBannerUseCase
        self.updateHomeBannerUseCase = updateHomeBannerUseCase
        self.getCafeManagementUseCase = getCafeManagementUseCase
        self.getHomeBannerManagementUseCase = getHomeBannerManagementUseCase
        self.getCafeNoticePageUseCase = getCafeNoticePageUseCase
        self.getCafeEventPageUseCase = getCafeEventPageUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.uploadImageUseCase = uploadImageUseCase

        observeSession()
        loadOwnedCafeOptions()
        loadEditingBanner()
    }

    deinit {
        selectorTask?.cancel()
    }

    private enum MessageKey {
        static let ownedCafeLoadFailed = "banneredit_info_owned_cafe_load_failed"
        static let editBannerNotFound = "banneredit_info_edit_banner_not_found"
        static let editBannerLoadFailed = "banneredit_info_edit_banner_load_failed"
        static let selectCafeFirst = "banneredit_info_select_cafe_first"
        static let noticeListLoadFailed = "banneredit_info_notice_list_load_failed"
        static let eventListLoadFailed = "banneredit_info_event_list_load_failed"
        static let validationTitleRequired = "banneredit_validation_title_required"
        static let validationImageRequired = "banneredit_validation_image_required"
        static let validationSubtitleRequired = "banneredit_validation_subtitle_required"
        static let validationExternalUrlRequired = "banneredit_validation_external_url_required"
        static let validationTargetRequired = "banneredit_validation_target_required"
        static let imageUploadFailed = "banneredit_info_image_upload_failed"
        static let unauthorized = "banneredit_error_unauthorized"
        static let permissionDenied = "banneredit_error_permission_denied"
        static let targetNotFound = "banneredit_error_target_not_found"
        static let networkFailed = "banneredit_error_network_failed"
        static let saveUnknown = "banneredit_error_save_unknown"
    }
}
